WifiART/src/userinterface/ZModel.java

```diff
@@ -62,9 +116,110 @@ public void setCallback(ZController fc) {
+            floorNumberEncoding = MAX_FLOOR_MAGNITUDE;
+        } else if (floorNumberEncoding < -MAX_FLOOR_MAGNITUDE) {
+            floorNumberEncoding = -MAX_FLOOR_MAGNITUDE;
+        }
+        if (floorNumber < 0) {
+            result |= FLOOR_SIGN_MASK; // Set the sign bit
+        }
+        // Place bits 0-12 of floorNumberEncoding into bits 2-14 of the result.
+        result |= ((floorNumberEncoding << FLOOR_INDEX) & FLOOR_MAGNITUDE_MASK);
+
+        return result;
+    }
+
+    private int getHeightAboveFloorField() {
+        int result = 0;
+
+        double heightAboveFloor = state.getHeightAboveFloor();
+        // Convert to 1/4096-ths of a meter, rounded.
+        long heightAboveFloorEncoding = Math.round(heightAboveFloor * HEIGHT_ABOVE_FLOOR_FRACTION_FACTOR);
+        if (heightAboveFloorEncoding > MAX_HEIGHT_ABOVE_FLOOR_MAGNITUDE) {
+            heightAboveFloorEncoding = MAX_HEIGHT_ABOVE_FLOOR_MAGNITUDE;
+        } else if (heightAboveFloorEncoding < -MAX_HEIGHT_ABOVE_FLOOR_MAGNITUDE) {
+            heightAboveFloorEncoding = -MAX_HEIGHT_ABOVE_FLOOR_MAGNITUDE;
+        }
+        if (heightAboveFloor < 0) {
+            result |= HEIGHT_SIGN_MASK; // Set the sign bit
+        }
+        result |= (heightAboveFloorEncoding & HEIGHT_MAGNITUDE_MASK); // Set the magnitude bits
+
+        return result;
+    }
+
+    private int getHeightAboveFloorUncertaintyField() {
+        int result = 0;
+
+        double heightAboveFloorUncertainty = state.getHeightAboveFloorUncertainty();
+        double heightAboveFloorUncertaintyEncoding;
+        if (isFractionBitsPresent) {
+            // Convert to 1/4096-ths of a meter.
+            heightAboveFloorUncertaintyEncoding = heightAboveFloorUncertainty * HEIGHT_ABOVE_FLOOR_FRACTION_FACTOR;
+        } else {
+            heightAboveFloorUncertaintyEncoding = heightAboveFloorUncertainty;
+        }
+        // encoding = 11 - log2(uncertainty)
+        heightAboveFloorUncertaintyEncoding = 11 - Math.log(heightAboveFloorUncertaintyEncoding) / Math.log(2);
+
+        // Convert to integer, rounding down so the bound still applies.
+        result = (int) (heightAboveFloorUncertaintyEncoding);
+        if (result < MIN_HEIGHT_ABOVE_FLOOR_UNCERTAINTY_ENCODING) {
+            result = MIN_HEIGHT_ABOVE_FLOOR_UNCERTAINTY_ENCODING;
+        } else if (result > MAX_HEIGHT_ABOVE_FLOOR_UNCERTAINTY_ENCODING) {
+            result = MAX_HEIGHT_ABOVE_FLOOR_UNCERTAINTY_ENCODING;
+        }
+
+        return result;
+    }
+
+    /**
+     * Insert an integer into a byte array in little-endian format.
+     *
+     * @param arr The byte array being populated
+     * @param num The integer to insert into the array
+     * @param startIndex The starting index that the integer should appear in the array
+     * @param length The number of bytes being populated
+     */
+    private void fillLittleEndian(byte[] arr, int num, int startIndex, int length) {
```

Comment:
This method is useful for general purpose of inserting an integer into a byte array. Could it be extracted to a utility class (do we have similar methods like this one)?

Reply:
Yes, it can be in a utility class. I'm not sure that this branch is the best place to change this though, since other subelements might use this and they are on different branches, so they wouldn't be able to access the utility class until this PR gets merged. I think this utility class will have to be in a separate PR.

Reply:
Sounds good. It's totally fine to refactor in a following up CL.